File: backend/app/api/v1/endpoints/bookings.py
"""
Booking Endpoints
API endpoints for bookings and reservations
"""

from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.logging import logger
from app.schemas.booking import BookingCreate, BookingResponse, BookingSummaryResponse
from app.services.booking_service import BookingService
from app.services.availability_service import AvailabilityService
from app.models.booking import Booking, BookingStatus
from app.models.masterclass import CityEvent
from sqlalchemy import select
from sqlalchemy.orm import selectinload

router = APIRouter()


@router.post("/create", response_model=BookingResponse, status_code=status.HTTP_201_CREATED)
async def create_booking(
    booking_data: BookingCreate,
    db: AsyncSession = Depends(get_db),
):
    """
    Create a new booking
    
    Args:
        booking_data: Booking creation data
        
    Returns:
        Created booking with reference
    """
    booking_service = BookingService(db)
    
    try:
        booking = await booking_service.create_booking(booking_data)
        
        # Reload with relationships
        result = await db.execute(
            select(Booking)
            .options(
                selectinload(Booking.attendees),
                selectinload(Booking.city_event).selectinload(CityEvent.event),
                selectinload(Booking.city_event).selectinload(CityEvent.city),
                selectinload(Booking.city_event).selectinload(CityEvent.venue),
            )
            .where(Booking.id == booking.id)
        )
        booking = result.scalar_one_or_none()
        
        return BookingResponse.model_validate(booking)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        logger.error(f"Error creating booking: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create booking",
        )


@router.get("/{booking_reference}", response_model=BookingResponse)
async def get_booking_by_reference(
    booking_reference: str,
    db: AsyncSession = Depends(get_db),
):
    """
    Get booking status by reference
    
    Args:
        booking_reference: Unique booking reference
        
    Returns:
        Booking details with status, attendees, and payment info
    """
    booking_service = BookingService(db)
    
    booking = await booking_service.get_booking_by_reference(booking_reference)
    
    if not booking:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Booking {booking_reference} not found",
        )
    
    return BookingResponse.model_validate(booking)


@router.post("/{booking_reference}/cancel", response_model=BookingResponse)
async def cancel_booking(
    booking_reference: str,
    db: AsyncSession = Depends(get_db),
):
    """
    Cancel a booking
    
    Args:
        booking_reference: Unique booking reference
        
    Returns:
        Cancelled booking details
    """
    booking_service = BookingService(db)
    
    try:
        booking = await booking_service.cancel_booking(booking_reference)
        
        # Reload with relationships
        result = await db.execute(
            select(Booking)
            .options(
                selectinload(Booking.attendees),
                selectinload(Booking.city_event).selectinload(CityEvent.event),
                selectinload(Booking.city_event).selectinload(CityEvent.city),
                selectinload(Booking.city_event).selectinload(CityEvent.venue),
            )
            .where(Booking.id == booking.id)
        )
        booking = result.scalar_one_or_none()
        
        return BookingResponse.model_validate(booking)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        logger.error(f"Error cancelling booking {booking_reference}: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to cancel booking",
        )


@router.get("/{booking_reference}/summary", response_model=BookingSummaryResponse)
async def get_booking_summary(
    booking_reference: str,
    db: AsyncSession = Depends(get_db),
):
    """
    Get booking summary for confirmation page
    
    Args:
        booking_reference: Unique booking reference
        
    Returns:
        Booking summary with city, venue, dates, and price
    """
    booking_service = BookingService(db)
    
    booking = await booking_service.get_booking_by_reference(booking_reference)
    
    if not booking:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Booking {booking_reference} not found",
        )
    
    # Load city_event with relationships
    result = await db.execute(
        select(CityEvent)
        .options(
            selectinload(CityEvent.city),
            selectinload(CityEvent.venue),
        )
        .where(CityEvent.id == booking.city_event_id)
    )
    city_event = result.scalar_one_or_none()
    
    if not city_event:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="City event not found",
        )
    
    return BookingSummaryResponse(
        booking_reference=booking.booking_reference,
        city=city_event.city.name_en,
        city_fr=city_event.city.name_fr,
        venue_name=city_event.venue.name,
        venue_address=city_event.venue.address,
        start_date=city_event.start_date.isoformat(),
        end_date=city_event.end_date.isoformat(),
        attendee_name=booking.attendee_name,
        attendee_email=booking.attendee_email,
        quantity=booking.quantity,
        total=booking.total,
        payment_status=booking.payment_status,
    )

File: apps/web/src/lib/api/bookings.ts
/**
 * Bookings API Client
 * 
 * Client for creating and managing bookings
 */

import { apiClient, extractApiData } from '@/lib/api';
import type { ApiResponse } from '@modele/types';

export interface AttendeeCreate {
  first_name: string;
  last_name: string;
  email: string;
  phone?: string;
  role?: string;
  experience_level?: string;
  dietary_restrictions?: string;
}

export interface BookingCreate {
  city_event_id: number;
  attendee_name: string;
  attendee_email: string;
  attendee_phone?: string;
  ticket_type: 'EARLY_BIRD' | 'REGULAR' | 'GROUP';
  quantity: number;
  attendees?: AttendeeCreate[];
}

export interface AttendeeResponse {
  id: number;
  booking_id: number;
  first_name: string;
  last_name: string;
  email: string;
  phone?: string;
  role?: string;
  experience_level?: string;
  dietary_restrictions?: string;
  created_at: string;
}

export interface BookingResponse {
  id: number;
  city_event_id: number;
  booking_reference: string;
  status: 'PENDING' | 'CONFIRMED' | 'CANCELLED';
  attendee_name: string;
  attendee_email: string;
  attendee_phone?: string;
  ticket_type: 'EARLY_BIRD' | 'REGULAR' | 'GROUP';
  quantity: number;
  subtotal: number;
  discount: number;
  total: number;
  payment_status: 'PENDING' | 'PAID' | 'FAILED' | 'REFUNDED';
  payment_intent_id?: string;
  created_at: string;
  confirmed_at?: string;
  cancelled_at?: string;
  attendees: AttendeeResponse[];
}

export interface BookingSummaryResponse {
  booking_reference: string;
  city: string;
  city_fr: string;
  venue_name: string;
  venue_address?: string;
  start_date: string;
  end_date: string;
  attendee_name: string;
  attendee_email: string;
  quantity: number;
  total: number;
  payment_status: 'PENDING' | 'PAID' | 'FAILED' | 'REFUNDED';
}

export interface PaymentIntentResponse {
  client_secret: string;
  payment_intent_id: string;
  amount: number;
  currency: string;
}

/**
 * Bookings API client
 */
export const bookingsAPI = {
  /**
   * Create a new booking
   */
  create: async (data: BookingCreate): Promise<BookingResponse> => {
    const response = await apiClient.post<BookingResponse>('/v1/bookings/create', data);
    const result = extractApiData<BookingResponse>(response as unknown as ApiResponse<BookingResponse>);
    if (!result) {
      throw new Error('Failed to create booking: no data returned');
    }
    return result;
  },

  /**
   * Get booking by reference
   */
  getByReference: async (reference: string): Promise<BookingResponse> => {
    const response = await apiClient.get<BookingResponse>(`/v1/bookings/${reference}`);
    const data = extractApiData<BookingResponse>(response as unknown as ApiResponse<BookingResponse>);
    if (!data) {
      throw new Error(`Booking not found: ${reference}`);
    }
    return data;
  },

  /**
   * Get booking summary by reference
   */
  getSummary: async (reference: string): Promise<BookingSummaryResponse> => {
    const response = await apiClient.get<BookingSummaryResponse>(`/v1/bookings/${reference}/summary`);
    const data = extractApiData<BookingSummaryResponse>(response as unknown as ApiResponse<BookingSummaryResponse>);
    if (!data) {
      throw new Error(`Booking summary not found: ${reference}`);
    }
    return data;
  },

  /**
   * Cancel a booking
   */
  cancel: async (reference: string): Promise<BookingResponse> => {
    const response = await apiClient.post<BookingResponse>(`/v1/bookings/${reference}/cancel`);
    const data = extractApiData<BookingResponse>(response as unknown as ApiResponse<BookingResponse>);
    if (!data) {
      throw new Error(`Failed to cancel booking: ${reference}`);
    }
    return data;
  },

  /**
   * Create payment intent for a booking
   */
  createPaymentIntent: async (bookingId: number): Promise<PaymentIntentResponse> => {
    const response = await apiClient.post<PaymentIntentResponse>(`/v1/bookings/${bookingId}/create-payment-intent`);
    const data = extractApiData<PaymentIntentResponse>(response as unknown as ApiResponse<PaymentIntentResponse>);
    if (!data) {
      throw new Error(`Failed to create payment intent for booking: ${bookingId}`);
    }
    return data;
  },
};
